### PR TITLE
record: Improve string representation

### PR DESCRIPTION
**Pull request by sakshamarora1**
Fixes: #612 
Fixes: #335 

```console
dffml predict all \
    -model tfdnnc \
    -model-batchsize 1000 \
    -model-hidden 30 50 25 \
    -model-clstype int \
    -model-predict label:int:1 \
    -model-classifications $(seq 0 9) \
    -model-features image:int:$((28 * 28)) \
    -sources images=csv \
    -source-filename image.csv  \
    -source-loadfiles image  \
    -pretty
    -log debug
```
Example output of 1 of the records:
```console
	Key:	four
                                                               Record Features
+----------------------------------------------------------------------------------------------------------------------------------------------+
|               image               |                 0,0,0,0,0,0,0,0,0,1,0,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,... (length:784)                 |
+----------------------------------------------------------------------------------------------------------------------------------------------+

                                                                  Prediction
+----------------------------------------------------------------------------------------------------------------------------------------------+
|                                                                    label                                                                     |
+----------------------------------------------------------------------------------------------------------------------------------------------+
|             Value:  3             |                                    Confidence:   0.13614192605018616                                     |
+----------------------------------------------------------------------------------------------------------------------------------------------+

```

For list records:
```console
dffml list records -sources f=csv -source-filename test.csv -pretty

	Key:	0
                                                               Record Features
+----------------------------------------------------------------------------------------------------------------------------------------------+
|             Expertise             |                                                    9                                                     |
+----------------------------------------------------------------------------------------------------------------------------------------------+
|               Salary              |                                                    50                                                    |
+----------------------------------------------------------------------------------------------------------------------------------------------+
|               Trust               |                                                   0.5                                                    |
+----------------------------------------------------------------------------------------------------------------------------------------------+
|               Years               |                                                    4                                                     |
+----------------------------------------------------------------------------------------------------------------------------------------------+
                                                                                                                   Prediction:    Undetermined

```

Signed-off-by: sakshamarora1 <sakshamarora1001@gmail.com>

### CODE REVIEW

**Comment by sakshamarora1** [comment]
@pdxjohnny I added numpy in record.py for dealing with ndarrays of images later that is why half the tests are failing.
I think you told me something about not wanting numpy as a requirement at the top level? Sorry, I don't remember completely.

If I remove checking for ndarray and and change the array to tuple/list after using the png configloader then for preprocessing operations I will have to do np.array(<input_array>) each time at the beginning. What do you think?

File: dffml/record.py [hunk 1]
```diff
@@ -3,6 +3,8 @@
 """
 Information on the software to evaluate is stored in a Record instance.
 """
+import os
+import numpy as np
```
Comment (pdxjohnny) [inline dffml/record.py:7]: Yes you were correct, we want to keep the top level package free of dependencies

```suggestion
```

File: dffml/record.py [hunk 2]
```diff
@@ -141,31 +143,75 @@ def __repr__(self):
...
+                        + "|"
+                        + (
+                            ", ".join(
+                                str(val) for val in results[: int(width / 10)]
+                            )
+                            + f",... (length:{len(results)})"
+                        ).center(int(3 * width / 4))
+                        + "|"
+                    )
+                    if isinstance(results, (list, tuple, np.ndarray,))
```
Comment (pdxjohnny) [inline dffml/record.py:175]:
```suggestion
                    if isinstance(results, (list, tuple,))
```

File: dffml/record.py [hunk 3]
```diff
@@ -141,31 +143,75 @@ def __repr__(self):
...
+                        + "|"
+                        + str(results).center(int(3 * width / 4))
+                        + "|"
+                    )
+                    if isinstance(results, (str, int, float, bool,))
+                    else ""
+                )
+                + "\n"
+                + divider
                 for feature, results in self.features().items()
```
[the remark this answers is not in the document]
Comment (pdxjohnny) [reply dffml/record.py:188]: If this suggestion results in an error then you'll want to go modify `export_value` in `dffml.util.data`

File: dffml/record.py [hunk 4]
```diff
@@ -141,31 +143,75 @@ def __repr__(self):
...
-            header += " " + str(self.extra)
+            header += "\n\t" + str(self.extra)
 
         return "\n".join(
             [header]
+            + [feature_header]
+            + [divider]
             + [
-                ("%-30s%s" % (feature, str(results)))
+                (
```
Comment (pdxjohnny) [inline dffml/record.py:162]: This "building a table from a dict" should be it's own function. Maybe it should go somewhere like `dffml.util.display`? `display` may not be the best word choice, if you think of something else go for it.

File: dffml/util/data.py [hunk 5]
```diff
@@ -215,6 +215,8 @@ def export_value(obj, key, value):
         obj[key] = value.export()
     elif hasattr(value, "_asdict"):
         obj[key] = value._asdict()
+    elif getattr(type(value), "__module__", None) == "numpy":
```
Comment (pdxjohnny) [inline dffml/util/data.py:218]: We should also check that the `flatten` method exists and can be called to be safe

```suggestion
    elif getattr(type(value), "__module__", None) == "numpy" and inspect.ismethod(getattr(value, "flatten", None)):
```

Comment (sakshamarora1) [reply dffml/util/data.py:218]: Alright!

**Pull request merged**